feat: add support for publishing workers with r2 bucket bindings

petebacondarwin opened the pull request:
This change adds the ability to define bindings in your `wrangler.toml` file
for R2 buckets. These buckets will then be available in the environment
passed to the worker at runtime.

Closes #365

threepointone commented on packages/wrangler/src/config.ts:
```diff
@@ -203,6 +203,15 @@ export type Config = {
 
+  r2_buckets?: {
+    /** The binding name used to refer to the R2 bucket in the worker. */
+    binding: string;
```
really dislike "binding" bah. I'll make some noise about it later.

threepointone reviewed (changes requested):
This looks good to me! Could you add it to `wrangler dev` as well? That would be needed to close the issue.

threepointone commented:
You could also add `wrangler dev` support in another PR, but then remove the "Closes" directive from the PR description, and I'll stamp this.

petebacondarwin commented:
That is so weird. I actually ran `wrangler dev` on my test project and I saw the R2 binding in the worker....
But looking through the code, again, I can see that it is not being added at the `<Dev>` component...
Is there any reason why dev extracts these bindings differently to how publish? Perhaps there is an opportunity for more de-duplication there?

petebacondarwin commented:
Looking through the code where `toFormData()` is used, it seems that we also need to update when submitting a secret. (Not quite sure why... is that related to the issue before of setting secrets where wiping out other environment variables?)

petebacondarwin commented:
Added the bindings to `wrangler dev` I'll check it manually later today.
I also changed the `CfWorkerInit` so that each of the `bindings` properties were required (but could still be undefined) so as to ensure we don't miss this again.
@threepointone - PTAL.

threepointone reviewed (approved):
![image](https://user-images.githubusercontent.com/18808/153745489-be7d83cd-57b4-4c10-92a5-b15187a2846e.gif)

threepointone commented:
> Is there any reason why dev extracts these bindings differently to how publish? Perhaps there is an opportunity for more de-duplication there?

Probably! It'll also force people to ship support for both whenever they make a new type of binding.
